plasmapy/atomic/ionization_states.py

```diff
@@ -0,0 +1,965 @@
+                 *,
+                 T_e=None,
+                 n_e=None,
+                 n_elem=None,
+                 tol: Union[float, int] = 1e-15):
+        """Initialize a `~plasmapy.atomic.IonizationState` instance."""
+
+        self._particle = particle
+
+        try:
+            self.tol = tol
+            self.T_e = T_e
+            self.n_elem = n_elem
+            self.n_e = n_e
+            self.ionic_fractions = ionic_fractions
+
+            # This functionality has not yet been implemented:
+
+            # if self._ionic_fractions is None and self.T_e is not None:
+            #     self.equilibrate()
+
+        except Exception as exc:
+            raise AtomicError(
+                f"Unable to create IonizationState instance for "
+                f"{particle.particle}.") from exc
+
+    def __getitem__(self, value) -> State:
+        """Return the ionic fraction(s)."""
+        if isinstance(value, slice):
+            raise TypeError("IonizationState instances cannot be sliced.")
+
+        if isinstance(value, (int, np.integer)) and 0 <= value <= self.atomic_number:
+            result = State(value, self.ionic_fractions[value], self.ionic_symbols[value])
+        else:
+            if not isinstance(value, Particle):
+                try:
+                    value = Particle(value)
+                except InvalidParticleError as exc:
+                    raise InvalidParticleError(
+                        f"{value} is not a valid integer charge or "
+                        f"particle.") from exc
+
+            same_element = value.element == self.element
+            same_isotope = value.isotope == self.isotope
+            has_charge_info = value.is_category(any_of=["charged", "uncharged"])
+
+            if same_element and same_isotope and has_charge_info:
+                Z = value.integer_charge
+                result = State(Z, self.ionic_fractions[Z], self.ionic_symbols[Z])
+            else:
+                if not same_element or not same_isotope:
+                    raise AtomicError("Inconsistent element or isotope.")
+                elif not has_charge_info:
+                    raise ChargeError("No integer charge provided.")
+        return result
+
+    def __iter__(self):
+        """Initialize the instance prior to an iteration."""
+        self._charge_index = 0
+        return self
+
+    def __next__(self):
+        """
+        Return a `~State` `~collections.namedtuple` which contains
+        ``integer_charge``, ``ionic_fraction``, and ``ionic_symbol``.
+        """
+        if self._charge_index <= self.atomic_number:
+            result = State(
+                self._charge_index,
+                self._ionic_fractions[self._charge_index],
+                self.ionic_symbols[self._charge_index],
+            )
+            self._charge_index += 1
+            return result
+        else:
+            del self._charge_index
+            raise StopIteration
```

Comment:
I'm just going through this file before going to bed today, so I'll finish the review later, but: this definitely needs a test. If it already has one, awesome.

... how many lines, again?!